vyper/venom/passes/memmerging.py

```diff
@@ -0,0 +1,274 @@
+            else:
+                i += 1
+
+        return True
+
+    def _overlap_exist(
+        self, intervals: list[_Interval], inter: _Interval, index: int | None = None
+    ) -> bool:
+        if index is None:
+            index = bisect_left(intervals, inter)
+
+        if index > 0:
+            if intervals[index - 1].overlap(inter):
+                return True
+
+        if index < len(intervals):
+            if intervals[index].overlap(inter):
+                return True
+
+        return False
+
+    def _handle_bb(
+        self, bb: IRBasicBlock, load_inst: str, copy_inst: str, allow_dst_overlaps_src: bool = False
+    ):
+        loads: dict[IRVariable, int] = dict()
+        intervals: list[_Interval] = []
+
+        def _barrier():
+            self._optimize_copy(bb, intervals, copy_inst)
+            loads.clear()
+
+        for inst in bb.instructions.copy():
+            if inst.opcode == load_inst:
+                src_op = inst.operands[0]
+                if not isinstance(src_op, IRLiteral):
+                    _barrier()
+                    continue
+                assert inst.output is not None  # help mypy
+                uses = self.dfg.get_uses(inst.output)
+                if len(uses) != 1:
+                    _barrier()
+                    continue
+                if uses.first().opcode != "mstore":
+                    continue
+                if self._overlap_exist(
+                    intervals, _Interval(src_op.value + 32, src_op.value, 32, [])
+                ):
+                    _barrier()
+                    continue
+                assert isinstance(inst.output, IRVariable)
+                loads[inst.output] = src_op.value
+            elif inst.opcode == "mstore":
+                var = inst.operands[0]
+                dst = inst.operands[1]
+                if not isinstance(dst, IRLiteral) or not isinstance(var, IRVariable):
+                    _barrier()
+                    continue
+                if var not in loads:
+                    _barrier()
+                    continue
+                src: int = loads[var]
+                mload_inst = self.dfg.get_producing_instruction(var)
+                assert mload_inst is not None  # help mypy
+                n_inter = _Interval(dst.value, src, 32, [mload_inst, inst])
+                if not self._add_interval(
+                    intervals, n_inter, allow_dst_overlap_src=allow_dst_overlaps_src
+                ):
+                    _barrier()
+            elif _volatile_memory(inst):
+                _barrier()
+
+        self._optimize_copy(bb, intervals, copy_inst)
+
+    def _add_zero_interval(self, intervals: list[_Interval], new_inter: _Interval) -> bool:
+        index = bisect_left(intervals, new_inter)
+        intervals.insert(index, new_inter)
+
+        i = max(index - 1, 0)
+        while i < min(index + 1, len(intervals) - 1):
+            merged = intervals[i].inclusive_merge(intervals[i + 1])
```

Comment:
why do we use `inclusive_merge()` for mzero merging but `merge()` for copy merging?